### ISSUE TITLE
get_Lstar() hangs randomly

### ISSUE THREAD

**Issue by wlwetzel** [spacepy/spacepy#351]
I'm trying to call get_Lstar() a bunch of times (specifically to get the L and MLT at various times and locations). I haven't had a problem doing this in the past, but I'm having issues with this set of data. 

Calling get_Lstar() causes my program to hang, whether I loop through all the loci or try to pass them as an array. What's really driving me crazy is that get_Lstar() isn't hanging on any one input, it hangs randomly, and sometimes even runs without issue. 

### Minimal example to reproduce issue:

```
import numpy as np
from spacepy.time import Ticktock
import spacepy.coordinates as spc
import spacepy.irbempy as irb
import pandas as pd

test_data = '/path/to/data.txt'
data = pd.read_csv(test_data,names=['alt','lat','lon'],index_col=0,parse_dates=True,sep='\t')
times = data.index.tolist()
ticks = Ticktock(times)
loci = np.stack((data['alt'],data['lat'],data['lon'] ),axis=1)
for i in range(len(times)):
    tick = Ticktock(times[i])
    coord = spc.Coords(loci[i] , 'RLL','sph')
    print(irb.get_Lstar(tick,coord))
```
Here is the data I'm using:
2013-01-12 08:11:04	1.1330560370753462	-80.577	266.768
2013-01-12 08:11:08	1.1330560370753462	-80.483	265.443
2013-01-12 11:31:30	1.1330403352041967	-80.303	266.584
2013-01-17 14:47:24	1.1327892010673364	-71.558	253.659
2013-01-17 14:47:28	1.1328048933583328	-71.764	253.284
2013-01-15 13:48:24	1.1328362875204785	-72.932	265.828
2013-01-15 13:48:28	1.1328362875204785	-73.133	265.396
2013-01-15 13:48:32	1.132851989391628	-73.333	264.953
2013-01-21 16:44:40	1.1325537592214723	-66.347	231.194
2013-01-21 16:44:44	1.1325694515124687	-66.565	230.955
2013-01-11 11:52:20	1.1330246429132005	-80.105	263.596
2013-01-10 07:12:54	1.133008950622204	-78.151	261.141
2013-01-10 07:12:58	1.1329932487510546	-77.991	260.296
2013-01-10 07:13:02	1.1329932487510546	-77.828	259.471
2013-01-14 12:29:44	1.1329775564600584	-77.785	270.471
2013-01-14 12:29:48	1.1329775564600584	-77.949	269.652
2014-01-09 21:46:36	1.1326793262899024	-67.476	154.778
2014-01-09 21:46:40	1.1326793262899024	-67.692	154.518
2014-01-09 21:46:44	1.132695028161052	-67.908	154.253
2014-01-09 21:46:48	1.132695028161052	-68.123	153.984

### OS, Python version, and dependency version information:

```
Linux-5.4.0-33-generic-x86_64-with-glibc2.29
sys.version_info(major=3, minor=8, micro=2, releaselevel='final', serial=0)
numpy=1.18.4
scipy=1.4.1
matplotlib=3.2.1
spacepy=0.2.1
```

**Comment by jtniehof** [comment]
Data file recreated with the tab-separation between columns except between date and time.
[data.txt](https://github.com/spacepy/spacepy/files/4753300/data.txt)

**Comment by jtniehof** [comment]
I was able to run this twelve times in a row without issue. Because you're using RLL, I wonder if you're being bitten by #213. This was fixed in #295 so if you're comfortable installing straight from git, you might want to give that a shot. We're working towards 0.2.2 but it's not going to happen, say, tomorrow.

I'd also suggest considering whether RLL is really the coordinate system your input data are in or if, say, spherical GEO makes sense.

**Comment by wlwetzel** [comment]
Is there any documentation for the coordinate systems used? The irbem manual doesn't seem to contain any, and the paper linked on the spacepy coords page does not cover all of those listed.

**Comment by wlwetzel** [comment]
Also, having changed the coordinate system it doesn't seem to be just an issue with RLL. With GEO coordinates it inevitably hangs as well, again on different lines in the data file every time.

**Comment by jtniehof** [comment]
Sadly the docs on that are fairly thin, particularly for things like RLL. I probably should sit down with the _Explanatory Supplement_ and F&H and try to make some at some point...usually I annoy @drsteve.

Thanks for trying the change of coordinates. I fell back to 0.2.1 and was able to run ten times successfully, so that's certainly not the whole story. Would you think ten runs would be sufficient to reproduce what you're seeing if my system were affected? I'm running each one in a new interpreter, so if that doesn't match what you were doing I should change that.

I can also set up a conda environment to exactly match your dependencies and see what effect that has, but it might take a bit.

Have you run `toolbox.update` to have the latest omni data? The failure you're showing is in getting the omni data, not the actual L* calculation.

Finally there's always the hand-wavey "have you run a memory check or similar?" It's always nice to blame somebody else's hardware.

And at POES altitudes I'd expect McIlwain L to be pretty reasonable for calculating conjunctions.

**Comment by drsteve** [comment]
## Hang on individual calls
I also can't reproduce the original error. I've tried several times, different coordinate systems... no luck.

## All-at-once error
I can reproduce the failure when running all-at-once. It's because the times are out of order. It should probably be documented (or trapped so we can raise an error), but `get_omni` currently requires ordered input times.

If you re-order your inputs
```
inds = np.argsort(times)
ticks = Ticktock(np.array(times)[inds])
loci = np.stack((data['alt'],data['lat'],data['lon'] ),axis=1)
cc = spc.Coords(loci, 'GEO', 'sph')
print(irb.get_Lstar(ticks, cc))
```
it should work fine (and does on my end).

## Coordinate system documentation
The IRBEM manual is definitely light on detail. They don't really define the coordinate systems, but they're mostly as defined in either the Hapgood papers on coordinate transforms or the appendix in Kivelson and Russell.

The Franz and Harper paper (linked from the Coords docstring) should cover everything except geodetic coordinates.

**Comment by wlwetzel** [comment]
### Hanging
Thanks for the ordering the inputs suggestion, that fixed the omni error, though now it just hangs like the loop did. 
This is seeming more like something on my end, as running that sample code ten times is always enough to reproduce the error. 
I did run a memory test, and everything looks fine, so I'm gonna try to track down exactly where in get_Lstar() the issue is. 

### Coordinate Systems
Thanks for the paper recommendations, those should be helpful.

**Comment by wlwetzel** [comment]
**Update**

I have now tried this on a different computer, with the same issues, so I don't think this is an issue with my hardware or setup. This is run on ubuntu 20.04. 

Something else I noticed was that toolbox.update() fails getting the leapseconds data. 

I also wrote a smaller script that doesn't require a data file:

`
import spacepy.coordinates as spc
import datetime
from spacepy.time import Ticktock
import spacepy.irbempy as irb
locus = [830.5  ,  55.428 ,  309.927]
coords = [locus] * 10
coords = spc.Coords(coords,'GDZ','sph')
times = [datetime.datetime(1994,1,1,5,10,i) for i in range(10)]
ticks = Ticktock(times)
lstar = irb.get_Lstar(ticks,coords,extMag='T89')

`
This will randomly hang about %20 of the time.

**Comment by jtniehof** [comment]
Thanks. I'll see if I can reproduce. Setting ncpus=1 works for you, correct? If so it's probably related to the multiprocessing setup and maybe #364.

As far as leap seconds, the USNO services have been down a LONG time and it took awhile to find an alternative source. See #338 and https://github.com/spacepy/spacepy/issues/233#issuecomment-554391251

**Comment by jtniehof** [comment]
I've run that example 100 times sequentially under Python 3.8, numpy 1.18.5, spacepy 0.2.1 and also spacepy master. This is using Anaconda on Ubuntu 16.04. I'll give a shot on an 18.04 machine and if I can't reproduce there, I'll have to start standing up a 20.04 VM (haven't had a chance to get a 20.04 machine running yet.)

**Comment by jtniehof** [comment]
No luck on 18.04. Can you clarify exactly how you're running this? (Script called from shell, calling from interactive Python session, using ipython?)

**Comment by wlwetzel** [comment]
I've tried calling it from shell and from ipython with similar results. I was on numpy 1.18.4, so I tried this same script with 1.18.5 and 1.19, with the same issue occurring. It does seem that in 1.19 and 1.18.5 the issue happens less frequently (1/100 times or so), though perhaps I was just unlucky when running 1.18.4.

**Comment by jtniehof** [comment]
Couldn't reproduce on 20.04, ran 200 times. I don't think I can do much more on this end. Maybe working #364 will give me some ideas (or get lucky and fix it as a side-effect.)

**Comment by wlwetzel** [comment]
Not really sure what I can do to debug, I'm not familiar with multiprocessing. For now I'll just use one cpu and I'll check again when #364 gets fixed.

**Comment by jtniehof** [comment]
Absolutely, there's not a lot you can do at this point. I'm going to leave this open because we have some weirdness that might be related and I'll loop back in once I've addressed #364. If nothing else I should be able to send you an instrumented drop-in to check. Also hope to have 0.2.2 out pretty soon and since that's a new irbem that might magically fix.

**Comment by jtniehof** [comment]
I'm marking this as can't reproduce, but also marking it as a blocker for 0.2.2. I can't promise we'll fix it for 0.2.2 but I'm not going to cut the release without revisiting it.

**Comment by jtniehof** [comment]
I might have a solution. Fire up ipython and do:
```
import spacepy.irbempy
print(spacepy.irbempy.__file__)
```

Unzip the attached and drop it in the same directory as the file printed from the above (you'll be overwriting `irbempy.py`, not the actual file listed.) Then try your test case again.
[irbem.zip](https://github.com/spacepy/spacepy/files/4892551/irbem.zip)

**Comment by wlwetzel** [comment]
This seems to work, I ran the test 100 times just fine.

**Comment by jtniehof** [comment]
Thanks for checking! The fix will be in 0.2.2, which I'm honestly trying to get out ASAP.
